### ISSUE TITLE
basic general purpose datasource

### ISSUE THREAD

**Issue by Dieterbe** [grafana/grafana#2965]
there's been some interest from projects such as https://github.com/twitter/diffy and i hear chef as well,
to be able to "just use grafana" to plot timeseries data that these processes have to offer.
one approach would be find the simplest tsdb (influxdb currently?), store the data in there, and query that, but it's still a hassle.
alternatively, we could have grafana pull the data straight out of those running processes.

perhaps we could specify a simple REST api, similar to graphite, to support use cases like this.
basically a datasource would be just a url.
grafana would query that url like `?from=<unix-ts>&to=<unix-ts>&target=<key1>[&target=<key2>]`
and then the process could return a json doc like
```
[
{ "target": "key",
"datapoints": [...]
}
]
```

**Comment by Dieterbe** [comment]
we could also have an http endpoint that lists all keys, to provide autocomplete in grafana.
but it should be ok for software not to implement this, to keep the bar as low as possible.
if that endpoint 404's we just gracefully fall back to no autocomplete, and you just have to manually specify the keys.

**Comment by torkelo** [comment]
yes, this would be great. Need to define a simple well documented http api standard, with possibly many levels of features (as you say optional features so the basic support is dead simple to implement)

**Comment by tdyas** [comment]
FYO, I am developing a work-in-progress JSON data source that can query a URL, substituting time values using $from and $to template variables . The results must be returned in the format expected as the result of Grafana's `query` method on the data source. See https://github.com/tdyas/grafana/tree/json_datasource.

Open issues:
- Each configured endpoint should support arbitrary substitution variables that would be selected in the metric editor.
- Supporting proxying through the Go backend. (I started working on this as you can see in the branch.) Currently, the endpoint must return the proper CORS header.
- The result format should be made configurable. Some additional thought is necessary to determine how to do this.

**Comment by addshore** [comment]
+1 for something like this

**Comment by Dieterbe** [comment]
@tdyas 
> The result format should be made configurable. Some additional thought is necessary to determine how to do this.

why? isn't something like json common enough? or what do you mean?

**Comment by addshore** [comment]
Well, what I would love from this datasource is to be able to configure a mapping of the JSON format returned from an API to the format required by grafana.

This means rather than reimplementing an existing api to output a different format specifically for this grafana plugin the change in format could instead me moved to the plugin and configuring of a datasource on a grafana install.

I started working on code for accessing a sprecific API with a specific format at https://phabricator.wikimedia.org/P2338

However if we could come up with a way to describe the relationship between one output format and the grafana expected format I may not need to work on this any more!

**Comment by tdyas** [comment]
So the initial version is in my company's Grafana fork here: https://github.com/foursquare/grafana/tree/v2.5.x_fs/public/app/plugins/datasource/json

I could move it out and submit to the grafana-plugins repo (https://github.com/grafana/grafana-plugins). Although some work in the server is needed to support proxying.

Maybe worth discussing on an email thread or a separate issue?

**Comment by sebastian-azeti** [comment]
Very interesting, but it seems that your plugin is not maintained anymore? Anyone knows if there is something else available to query standard HTTP APIs through grafana?

**Comment by tdyas** [comment]
It isn't. The reason why I originally developed the plugin for my company went away.
